private/model/api/passes.go

```diff
@@ -124,15 +124,27 @@ func (a *API) renameToplevelShapes() {
 	for _, v := range a.Operations {
 		if v.HasInput() {
 			name := v.ExportedName + "Input"
-			switch n := len(v.InputRef.Shape.refs); {
-			case n == 1 && a.Shapes[name] == nil:
+			switch {
+			case a.Shapes[name] == nil:
+				service, ok := shamelist[a.name]
+				if ok {
+					if check, ok := service[v.Name]; ok && check.input {
+						break
+					}
+				}
 				v.InputRef.Shape.Rename(name)
 			}
 		}
 		if v.HasOutput() {
 			name := v.ExportedName + "Output"
-			switch n := len(v.OutputRef.Shape.refs); {
-			case n == 1 && a.Shapes[name] == nil:
+			switch {
+			case a.Shapes[name] == nil:
+				service, ok := shamelist[a.name]
+				if ok {
```

Comment:
suggest merging `ok` if check up to the assignment above since both value are used within the scope of the if